applications/FluidDynamicsApplication/custom_elements/surface_smoothing_element.cpp

```diff
@@ -0,0 +1,670 @@
+
+    for(unsigned int i=0; i<num_nodes; i++){
+        rResult[i]  =  this->GetGeometry()[i].GetDof(DISTANCE).EquationId();
+    }
+}
+
+/**
+ * determines the elemental list of DOFs
+ * @param ElementalDofList: the list of DOFs
+ * @param rCurrentProcessInfo: the current process info instance
+ */
+void SurfaceSmoothingElement::GetDofList(DofsVectorType& rElementalDofList, ProcessInfo& CurrentProcessInfo)
+{
+    // Here, num_dof = num_nodes
+    if (rElementalDofList.size() != num_nodes){
+        rElementalDofList.resize(num_nodes);
+    }
+
+    for(unsigned int i=0; i<num_nodes; i++){
+        rElementalDofList[i] = this->GetGeometry()[i].pGetDof(DISTANCE);
+    }
+}
+
+/**
+ * ELEMENTS inherited from this class have to implement next
+ * CalculateLocalSystem, CalculateLeftHandSide and CalculateRightHandSide methods
+ * they can be managed internally with a private method to do the same calculations
+ * only once: MANDATORY
+ */
+
+/**
+ * this is called during the assembling process in order
+ * to calculate all elemental contributions to the global system
+ * matrix and the right hand side
+ * @param rLeftHandSideMatrix: the elemental left hand side matrix
+ * @param rRightHandSideVector: the elemental right hand side
+ * @param rCurrentProcessInfo: the current process info instance
+ */
+void SurfaceSmoothingElement::CalculateLocalSystem(
+    MatrixType& rLeftHandSideMatrix,
+    VectorType& rRightHandSideVector,
+    ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    const double dt = rCurrentProcessInfo.GetValue(DELTA_TIME);
+    const double he = ElementSizeCalculator<3,4>::AverageElementSize(GetGeometry());
+    const double epsilon = 1.0e2*dt*he*he;
+
+    BoundedMatrix<double,num_nodes,num_dim> DN_DX;  // Gradients matrix
+    array_1d<double,num_nodes> N; // dimension = number of nodes . Position of the gauss point
+
+    array_1d<double,num_nodes> PHIdof; //dimension = number of DOFs, needed since we are using a residualbased approach
+    array_1d<double,num_nodes> PHIold; //dimension = number of DOFs
+    array_1d<VectorType,num_nodes> GradPHIold;
+
+    BoundedMatrix<double,num_nodes,num_nodes> tempM;
+    tempM = ZeroMatrix(num_nodes,num_nodes);
+    BoundedMatrix<double,num_nodes,num_nodes> tempA;
+    tempA = ZeroMatrix(num_nodes,num_nodes);
+
+    array_1d<double,num_nodes> tempBCRHS = ZeroVector(num_nodes);
+
+    if(rLeftHandSideMatrix.size1() != num_nodes) // be carefull that num_dof = num_nodes
+        rLeftHandSideMatrix.resize(num_nodes,num_nodes,false); //resizing the system in case it does not have the right size
+
+    if(rRightHandSideVector.size() != num_nodes)
+        rRightHandSideVector.resize(num_nodes,false);
+
+    // Getting data for the given geometry
+    double volume;
+    GeometryUtils::CalculateGeometryData(GetGeometry(), DN_DX, N, volume); //asking for gradients and other info
+
+    for(unsigned int i = 0; i<num_nodes; i++){
+        PHIold[i] = GetGeometry()[i].GetValue(DISTANCE);
```

Comment:
Are you using the non-historical database to save the old `DISTANCE` value? If this old value is the previous time step one, you can access it with the buffer position one of the historical database by doing `GetGeometry()[i].FastGetSolutionStepValue(DISTANCE, 1);`.

Reply:
Yes. Is there any effect on the performance? (comparing historical-buffer and non-historical)

Reply:
the historical is faster but requites more memory (bcs it is historical). Plus one needs to know beforehand such that it can be allocated with `AddNodalSolutionStepVariable`

Reply:
So, I will use the buffer position one for phi_old.